crates/ra_assists/src/handlers/fill_match_arms.rs

```diff
@@ -56,31 +50,115 @@ pub(crate) fn fill_match_arms(ctx: AssistCtx) -> Option<Assist> {
     }
 
     let db = ctx.db;
-
     ctx.add_assist(AssistId("fill_match_arms"), "Fill match arms", |edit| {
-        let indent_level = IndentLevel::from_node(match_arm_list.syntax());
+        let mut arms: Vec<MatchArm> = match_arm_list.arms().collect();
+        if arms.len() == 1 {
+            if let Some(Pat::PlaceholderPat(..)) = arms[0].pat() {
+                arms.clear();
+            }
+        }
 
-        let new_arm_list = {
-            let arms = variants
-                .into_iter()
-                .filter_map(|variant| build_pat(db, module, variant))
-                .map(|pat| make::match_arm(iter::once(pat), make::expr_unit()));
-            indent_level.increase_indent(make::match_arm_list(arms))
-        };
+        let mut has_partial_match = false;
+        let variants: Vec<MatchArm> = variants
+            .into_iter()
+            .filter_map(|variant| build_pat(db, module, variant))
+            .filter(|variant_pat| {
+                !arms.iter().filter_map(|arm| arm.pat().map(|_| arm)).any(|arm| {
```

Comment:
> !arms.iter().filter_map(|arm| arm.pat().map(|_| arm)).any(|arm| {

That looks like a great candidate for the refactoring.

Tbh, I find it a bit hard to understand what does the whole thing checks for, so we can extract it into a method and name it properly.

Then the `.filter_map(|arm| arm.pat().map(|_| arm)).any(|arm| {` might be also replaced with something like `arms.iter().filter_map(|arm| Some((arm.pat()?, arm.guard())))` to avoid unwraps and 
funny `.map(|_| arm))` things.

Finally, it looks like an easy change to get rid of the leading `!`, but that'd be better evaulated after the rest of the refactoring done.